mlir/include/mlir/IR/OpImplementation.h

```diff
@@ -21,6 +21,19 @@
 #include "llvm/Support/SMLoc.h"
 #include <optional>
 
+namespace {
+// reference https://stackoverflow.com/a/16000226
+template <typename T, typename = void>
+struct HasStaticName : std::false_type {};
+
+template <typename T>
+struct HasStaticName<T,
```

Comment:
As an alternative to this, can we enforce that each type has a `getTypeName` method? Similar to how each operation class must have a `getOperationName` method?